src/Analyzers/Analyzers/src/CompilationFeatureDetector.cs

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Analyzers
+{
+    internal static class CompilationFeatureDetector
+    {
+        public static async Task<IImmutableSet<string>> DetectFeaturesAsync(
+            Compilation compilation,
+            CancellationToken cancellationToken = default)
+        {
+            var symbols = new StartupSymbols(compilation);
+            if (!symbols.HasRequiredSymbols)
+            {
+                // Cannot find ASP.NET Core types.
+                return ImmutableHashSet<string>.Empty;
+            }
+
+            var features = ImmutableHashSet.CreateBuilder<string>();
+
+            // Find configure methods in the project's assembly
+            var configureMethods = ConfigureMethodVisitor.FindConfigureMethods(symbols, compilation.Assembly);
+            for (var i = 0; i < configureMethods.Count; i++)
+            {
+                var configureMethod = configureMethods[i];
+
+                // Handles the case where a method is using partial definitions. We don't expect this to occur, but still handle it correctly.
+                var syntaxReferences = configureMethod.DeclaringSyntaxReferences;
+                for (var j = 0; j < syntaxReferences.Length; j++)
+                {
+                    var semanticModel = compilation.GetSemanticModel(syntaxReferences[j].SyntaxTree);
+
+                    var syntax = await syntaxReferences[j].GetSyntaxAsync(cancellationToken).ConfigureAwait(false);
+                    var operation = semanticModel.GetOperation(syntax);
```

Comment:
More for my benefit, but is there a reason to prefer using operation over just looking for `InvocationExpressionSyntax`? You can effectively get to the Symbol using the latter.

Reply:
Operation is the cool new thing. It's a little more high level, and includes the merge of symbols + syntax. Operation is also language independent.